packages/tests/server/react/useQueries.test.tsx

```diff
@@ -0,0 +1,58 @@
+import { getServerAndReactClient } from './__reactHelpers';
+import { GetOptions } from '@tanstack/react-query/build/lib/useQueries';
+import { render, waitFor } from '@testing-library/react';
+import { TRPCClientError } from '@trpc/client';
+import { initTRPC } from '@trpc/server/src';
+import { konn } from 'konn';
+import React from 'react';
+import { z } from 'zod';
+
+const ctx = konn()
+  .beforeEach(() => {
+    const t = initTRPC.create({
+      errorFormatter({ shape }) {
+        return {
+          ...shape,
+          data: {
+            ...shape.data,
+            foo: 'bar' as const,
+          },
+        };
+      },
+    });
+    const appRouter = t.router({
+      post: t.router({
+        byId: t.procedure
+          .input(
+            z.object({
+              id: z.string(),
+            }),
+          )
+          .query(() => '__result' as const),
+      }),
+    });
+
+    return getServerAndReactClient(appRouter);
+  })
+  .afterEach(async (ctx) => {
+    await ctx?.close?.();
+  })
+  .done();
+
+test('useQueries()', async () => {
+  const { proxy, App } = ctx;
+  function MyComponent() {
+    const results = proxy.useQueries((t) => [t.post.byId({ id: '1' })]);
```

Comment:
This test is failing because of this error:
```ts
Type 'TrpcQueryOptionsForUseQueries<"post.byId", { id: string; }, "__result", unknown, TRPCClientError<CreateRouterInner<$Config, { byId: BuildProcedure<"query", { _config: $Config; _meta: {}; _ctx_out: {}; _input_in: { id: string; }; _input_out: { id: string; }; _output_in: unique symbol; _output_out: unique symbol; }, "...' is not assignable to type 'UseQueryOptionsForUseQueries<"__result", unknown, "__result", ["post.byId", { id: string; }]>'.
Types of property 'onError' are incompatible.
Type '((err: TRPCClientError<CreateRouterInner<$Config, { byId: BuildProcedure<"query", { _config: $Config; _meta: {}; _ctx_out: {}; _input_in: { id: string; }; _input_out: { id: string; }; _output_in: unique symbol; _output_out: unique symbol; }, "__result">; }>>) => void) | undefined' is not assignable to type '((err: unknown) => void) | undefined'.
Type '(err: TRPCClientError<CreateRouterInner<$Config, { byId: BuildProcedure<"query", { _config: $Config; _meta: {}; _ctx_out: {}; _input_in: { id: string; }; _input_out: { id: string; }; _output_in: unique symbol; _output_out: unique symbol; }, "__result">; }>>) => void' is not assignable to type '(err: unknown) => void'.
Types of parameters 'err' and 'err' are incompatible.
Type 'unknown' is not assignable to type 'TRPCClientError<CreateRouterInner<$Config, { byId: BuildProcedure<"query", { _config: $Config; _meta: {}; _ctx_out: {}; _input_in: { id: string; }; _input_out: { id: string; }; _output_in: unique symbol; _output_out: unique symbol; }, "__result">; }>>'.ts(2322)
```
cc @TkDodo this feels like an error in the `QueriesOptions` type for React Query. `t.post.byId({ id: '1' })` is (essentially) typed as `UseQueryOptions<inferProcedureOutput<TProcedure>, TRPCClientError<TRouter>, TData, [TPath, inferProcedureInput<TProcedure>]>` so I don't know why this is happening unless the underlying inference is incorrect. I don't see `QueriesOptions` checking for the `UseQueryOptions` type at all anywhere so that could be the issue too.

Reply:
The typings of `useQueries` are a total mystery to me 😅

Reply:
😱

Reply:
@artysidorenko maybe 😅 ?

Reply:
Hi folks,

Just looking at the type def for trpc `UseQueries`

```typescript
export type UseQueries<TRouter extends AnyRouter> = <T extends any[]>(
  queriesCallback: (
    t: UseQueriesProcedureRecord<TRouter>,
  ) => readonly [...QueriesOptions<T>],
  context?: UseQueryOptions['context'],
) => QueriesResults<T>;
```

Am I right in thinking `t.post.byId({ id: '1' })` should be the result of the queryFn? If so, that won't fit into `QueriesOptions<T>` which is expecting an options object. I think in your test example the type would need to look something like `(t) => [{queryFn: t.post.byId({ id: '1' })}]` to fit, if I understood correctly.

I think getting the order of options and results inferred correctly (the `[Promise1, Promise2]` vs `Promise[]` piece mentioned [here](https://github.com/trpc/trpc/issues/3188#issuecomment-1327416433)) is going to be a challenge - for context, the exported `QueriesOptions` helper you're using is so bulky because it's manually unwrapping the array of options object types and inferring what each individual element looks like, then `QueriesResults` is manually wrapping it back into a new array of result types (to preserve the strict order). If you're now trying to infer it from the result of the callback `queriesCallback`, it's going to get really messy :flushed:

One alternative that might be worth exploring for the types is to use an overloaded function definition instead. You'd need to write out lots of overloads to handle longer tuples, but it might be a bit more readable/maintainable?

<details>

<summary> something like this (example only goes up to 2) </summary>

```typescript
// handle 1 element tuple
function useQueries<TRouter extends AnyRouter, TQueryFnData1 = unknown>(
  queriesCallback:(
    t: UseQueriesProcedureRecord<TRouter>,
  ) => readonly [UseQueryOptions<TQueryFnData1>],
  context?: UseQueryOptions['context']
): [UseQueryResult<TQueryFnData1>]
// handle 2 element tuple - preserve variadic types
function useQueries<TRouter extends AnyRouter, TQueryFnData1 = unknown, TQueryFnData2 = unknown>(
  queriesCallback:(
    t: UseQueriesProcedureRecord<TRouter>,
  ) => readonly [UseQueryOptions<TQueryFnData1>,UseQueryOptions<TQueryFnData2>],
  context?: UseQueryOptions['context']
): [UseQueryResult<TQueryFnData1>, UseQueryResult<TQueryFnData2>]
// ^ add more overloads to handle bigger tuples

// default case - homogenous array of any length
function useQueries<TRouter extends AnyRouter, TQueryFnData = unknown>(
  queriesCallback:(
    t: UseQueriesProcedureRecord<TRouter>,
  ) => readonly UseQueryOptions<TQueryFnData>[],
  context?: UseQueryOptions['context']
): UseQueryResult<TQueryFnData>[]
{
  // actual function implementation
}
```

</details>

Reply:
Thanks for the detailed response!

> Am I right in thinking t.post.byId({ id: '1' }) should be the result of the queryFn?

No, that function returns a type matching `UseQueryOptions`:
https://github.com/trpc/trpc/blob/d5d5e5adf59aa7933bb12e6a37eab0d83e25bf64/packages/react-query/src/shared/proxy/useQueriesProxy.ts#L33-L39.

> I think getting the order of options and results inferred correctly (the [Promise1, Promise2] vs Promise[] piece mentioned here) is going to be a challenge - for context, the exported QueriesOptions helper you're using is so bulky because it's manually unwrapping the array of options object types and inferring what each individual element looks like, then QueriesResults is manually wrapping it back into a new array of result types (to preserve the strict order). If you're now trying to infer it from the result of the callback queriesCallback, it's going to get really messy 😳

That's true but I think that should be handled by React Query instead. Inferring what the individual elements look like is also what's not working here. The type returned from t.post.byId() is UseQueryOptions but it fails to infer it and expects `err` to be `unknown` instead of what I passed. And looking into the `GetOptions` type, I don't see it handling the inference for certain things properly either. It doesn't capture errors at all unless I pass an `error` property (which is not a valid option?) and even if I do have an error property, the TQueryKey and possibly TData inference doesn't work.

Ideally it could just do `UseQueryOptions<infer TQueryFnData, infer Error, infer TData, infer QueryKey> but it doesn't and I'm not sure why exactly.

Reply:
> No, that function returns a type matching UseQueryOptions

Ah ok, gotcha, I misread the API!

In that case yeah, your queryCallback has the explicit `TError = TRPCClientError<TRouter>`, but on the `UseQueries` end we struggle to infer `TError` from anywhere in the options so it's stuck with `unknown` (a bit more context at the end of the message).

If you need an mvp solution/to clean up the tests, for now - you can try the below diff screenshot which should at least maintain all the _results_ inference correctly, but it means your error type will be left as `unknown`.

I will have more of a think as well if there's a way to fix the error type when you _know_ that it's always gunna be `TRPCClientError<TRouter>`.

<details>

<summary> diff screenshot </summary>

Line 18 => leave TError as unknown
Line 24 (separate issue) => TData should default to TFnData when you don't have any selector in your query options

![image](https://user-images.githubusercontent.com/26905074/205521283-25e3355a-3238-4b7a-b53a-f9c8b920392f.png)

</details>


> Ideally it could just do `UseQueryOptions<infer TQueryFnData, infer Error, infer TData, infer QueryKey>` but it doesn't and I'm not sure why exactly

You mean, we unwrap the array and try to infer each element using `UseQueryOptions` directly? This was my first port of call back in the day, but because `UseQueryOptions` has these type params appearing multiple times as context-sensitive types - e.g. TData both a callback param and a func return type - TS runs into problems handling both the array-logic and the contextual inference.. why you might see it's inferring from object literal signatures instead (going slightly off-topic 😅 ... let me know if you find something different/better though, been looking for ways to improve it 🙏 )...